src/fd_cache.nim

```diff
@@ -0,0 +1,310 @@
+##
+## Copyright (c) 2024, Crash Override, Inc.
+##
+## This file is part of Chalk
+## (see https://crashoverride.com/docs/chalk)
+##
+
+##
+## FD cache
+## Global cache for opened files
+##
+## The overall concept is that there is a global cache
+## for all opened file streams.
+##
+## Some things you can do:
+## * yield   - create or get existing file stream from cache.
+##             Analogy is getting connection from pool.
+## * release - release file stream back into the cache.
+##             Analogy is release connection back into pool.
+## * close   - close file stream and delete from the cache.
+##             Analogy is closing connection.
+##             This should be used only in specific places where
+##             FD needs to be explicitly closed such as when overwriting
+##             a content of a file.
+##
+## To fascilitate above operations, the cache is
+## * limited to a specific size
+## * it keeps track of all the users of a specific file stream
+## * when cache reaches its limit size, it closes LRU file streams
+##
+
+# TODO move this to nimutils
+
+import std/[enumerate, posix, streams, tables]
+import nimutils/file
+
+# ----------------------------------------------------------------------------
+
+proc getOpenLimit(): int =
```

Comment:
It may be worth limiting ourselves to some percentage of the rlimit (or subtract out a healthy number). Hard to know what 3rd party stuff we'll pull in that can't use the API.

Reply:
yep. we already do. we still honor the `cache_fd_limit` config:

https://github.com/crashappsec/chalk/blob/caf6e274b6ed384e476511fd16a898a7245bbbd1/src/util.nim#L301

and it checks that value is less than 1/2 of the rlimit:

https://github.com/crashappsec/chalk/blob/caf6e274b6ed384e476511fd16a898a7245bbbd1/src/fd_cache.nim#L223-L234

Reply:
also for context default fd limit in the config is 50:

https://github.com/crashappsec/chalk/blob/938e4746a852dbfbf75337f8b82451bc98c22af6/src/configs/chalk.c42spec#L2524-L2527

so we should be ok for most systems. dont imagine well see anything much lower with rlimit of 1024